//
// PSMoveProtocol.proto: protobuf definition file for the PSMove Service Protocol
//
// Brendan Walker (brendan@millerwalker.net)
//
syntax = "proto3";
package PSMoveProtocol;

// Reliable (TCP) requests that can issued to the PSMove Service
message Request {
    // Unique id for this request. 
    // Used to keep track of which response is associated with which request
    // when multiple requests are in-flight.
    int32 request_id = 1;

    // The request type determines which of the request structures is actually used
    enum RequestType {
        START_PSMOVE_DATA_STREAM = 0;
        STOP_PSMOVE_DATA_STREAM = 1;
        SET_RUMBLE = 2;
        RESET_POSE = 3;
    }
    RequestType type = 2;
   
    // Parameters for START_PSMOVE_DATA_STREAM
    // NOTE: ControllerDataFrame packets will start streaming to client upon receiving this request
    message RequestStartPSMoveDataStream {
        int32 psmove_id = 1;
    }
    RequestStartPSMoveDataStream request_start_psmove_data_stream = 20;

    // Parameters for STOP_PSMOVE_DATA_STREAM
    // NOTE: ControllerDataFrame packets will stop streaming to client upon receiving this request
    message RequestStopPSMoveDataStream {
        int32 psmove_id = 1;
    }
    RequestStopPSMoveDataStream request_stop_psmove_data_stream = 21;

    // Parameters for SET_RUMBLE
    message RequestSetRumble {
        int32 psmove_id = 1;
        int32 rumble = 2; // [0,255]
    }
    RequestSetRumble request_rumble = 22;

    // Parameters for RESET_POSE
    message RequestResetPose {
        int32 psmove_id = 1;
    }
    RequestResetPose reset_pose = 23;   
}

// Reliable (TCP) responses to requests
message Response {
    // The response type
    enum ResponseType {
        GENERAL_RESULT= 0;
        CONNECTION_INFO= 1;
    }

    enum ResultCode {
        RESULT_OK= 0;
        RESULT_ERROR= 1;
        RESULT_CANCELED= 2;
    }

    ResponseType type = 1;
    int32 request_id = 2; // the request id that spawned this response, -1 if this is a notification
    ResultCode result_code = 3;
    
    // No Parameters for GENERAL_RESULT
    
    // Parameters for CONNECTION_INFO
    // This is returned automatically when connecting via TCP
    message ResultConnectionInfo {
        int32 tcp_connection_id = 1;
    }
    ResultConnectionInfo result_connection_info = 20;
}

// Unreliable (UDP) controller data packet format
message ControllerDataFrame
{
    // The id of controller this data is for
    int32 psmove_id= 1;
    
    // Monotonically increasing number.
    // Used to throw out old data if it arrives out of order.
    int32 sequence_num= 2;

    // Controller status flags
    bool IsConnected= 3;
    bool IsTrackingEnabled= 4;
    bool IsCurrentlyTracking= 5;

    // Tracking position in the space of the HMD tracking camera
    message Position {
        float x = 1;
        float y = 2;
        float z = 3;
    }
    Position position = 6;

    // Tracking orientation in the frame of the local magnetic field
    message Orientation{
        float x = 1;
        float y = 2;
        float z = 3;
        float w = 4;
    }
    Orientation orientation = 7;

    // Button Index Bits
    enum ButtonType {
        TRIANGLE= 0;
        CIRCLE= 1;
        CROSS= 2;
        SQUARE= 3;
        SELECT= 4;
        START= 5;
        PS= 6;
        MOVE= 7;
        TRIGGER= 8;
    }
    
    // Raw bitmask of which buttons are currently down
    // Buttons bits are indexed using the ButtonType enum
    uint32 button_down_bitmask = 8;

    // Trigger analog value [0,255]
    int32 trigger_value = 9;
}
